chapters/VK_EXT_present_timing/queries.txt

```diff
@@ -0,0 +1,441 @@
+available some time later.
+These time values can be asynchronously queried, and will be returned if
+available.
+All time values are in nanoseconds, according to the time-domain being used.
+
+To asynchronously query the presentation engine, for newly-available timing
+information about one or more previous presents to a given swapchain, call:
+
+include::{generated}/api/protos/vkGetPastPresentationTimingEXT.txt[]
+
+  * pname:device is the device associated with pname:swapchain.
+  * pname:swapchain is the swapchain to obtain presentation timing
+    information duration for.
+  * pname:pPresentationTimingCount is a pointer to an integer related to the
+    number of sname:VkPastPresentationTimingEXT structures to query, as
+    described below.
+  * pname:pPresentationTimings is either `NULL` or a pointer to an an array
+    of sname:VkPastPresentationTimingEXT structures.
+
+If pname:pPresentationTimings is `NULL`, then the number of newly-available
+timing records for the given pname:swapchain is returned in
+pname:pPresentationTimingCount.
+Otherwise, pname:pPresentationTimingCount must: point to a variable set by
+the user to the number of elements in the pname:pPresentationTimings array,
+and on return the variable is overwritten with the number of structures
+actually written to pname:pPresentationTimings.
+If the value of pname:pPresentationTimingCount is less than the number of
+newly-available timing records, at most pname:pPresentationTimingCount
+structures will be written.
+If pname:pPresentationTimingCount is smaller than the number of
+newly-available timing records for the given pname:swapchain,
+ename:VK_INCOMPLETE will be returned instead of ename:VK_SUCCESS to indicate
+that not all the available values were returned.
+
+include::{generated}/validity/protos/vkGetPastPresentationTimingEXT.txt[]
+--
+
+[open,refpage='VkPastPresentationTimingEXT',desc='Structure containing timing information about a previously-presented image',type='structs']
+--
+
+The sname:VkPastPresentationTimingEXT structure is defined as:
+
+include::{generated}/api/structs/VkPastPresentationTimingEXT.txt[]
+
+  * pname:sType is the type of this structure.
+  * pname:pNext is `NULL` or a pointer to an extension-specific structure.
+  * pname:presentID is an application-provided value that was given to a
+    previous fname:vkQueuePresentKHR command via
+    slink:VkPresentTimeEXT::pname:presentID.
+    It can: be used to uniquely identify a previous present with the
+    flink:vkQueuePresentKHR command.
+  * pname:targetPresentTime is an application-provided value that was given
+    to a previous flink:vkQueuePresentKHR command via
+    slink:VkPresentTimeEXT::pname:targetPresentTime.
+    If non-zero, it was used by the application to indicate that an image
+    not be presented any sooner than pname:targetPresentTime.
+  * pname:actualPresentTime is the time when the image of the
+    pname:swapchain was actually displayed.
```

Comment:
In addition to @kleinerm's [suggestion](https://github.com/KhronosGroup/Vulkan-Docs/pull/1364#discussion_r528006423) to provide timing variability, I would like to see actualPresentTime a bit more rigorously defined.  Does "displayed" refer to when the display is illuminated after scanout completes?  Is this the precise start, end or midpoint of that period?  What about rolling displays which illuminate a band of pixels as they are being scanned out?

I propose reporting the start of vsync instead.  Then with the ability to query more detailed timing info (e.g. vblank, front porch, clock rate) any other specific required relative points of time can be calculated.  This would also enable reporting the time prior to when actual illumination takes place.

Reply:
I agree with this. While it would be nice to know when a present hits the users eyeballs the start of vsync to that point in time is usually ~constant and if the system knows it could be communicated as a separate field.

Reply:
I propose end of vblank / aka start of scanout / when the first top-left pixel of the new image leaves the video output of the gpu as the point in time, because that is what is used by any other extension i am aware of (e.g, on Linux VDPAU, GLX/OML_sync_control, X11/DRI3/Present, X11/INTEL_swap_events, Wayland presentation_feedback, kernel DRM/KMS vblank and pageflip events, on Apple macOS/iOS Metal drawable presentation timestamps) etc. and also the VK_GOOGLE_display_timing extensions.

This is consistent with all past efforts, and easy to implement correctly. Especially useful if one wants to implement other past extensions on top of this one.

Should also work with VRR displays, at least as far as our current Linux/DRM/KMS implementation for AMD's amdgpu-kms driver goes, and some experiments i did on NVidia G-Sync hardware against the proprietary driver suggest the same approach works there.